stock_available_to_promise_release/models/stock_move.py

```diff
@@ -87,34 +88,36 @@ def _is_unreleaseable(self):
             and self.rule_id.available_to_promise_defer_pull
         )
 
+    def _has_unreleasable_state(self):
+        self.ensure_one()
+        if self.rule_id.allow_unrelease_return_done_move:
+            blocking_states = ("cancel",)
+        else:
+            blocking_states = ("done", "cancel")
+        return self.state not in blocking_states
+
     def _in_progress_for_unrelease(self) -> StockMoveBase:
         """
-        This method will return the moves not done or canceled that :
+        This method will return the moves with unreleasable state that :
 
         - have their picking printed
-        - have a quantity done != 0
-
+        - have a quantity done set if allow_unrelease_return_done_move
         """
-        moves = self.filtered(lambda m: m.state not in ("done", "cancel"))
-        if not moves:
-            return moves
-        moves_printed = moves.filtered("picking_id.printed")
-        if moves_printed:
-            return moves_printed
-        moves_done = moves.filtered("quantity_done")
-        if moves_done:
-            return moves_done
-        return moves.browse()
+        unreleasable_moves = self.filtered(lambda m: m._has_unreleasable_state())
+        if any(unreleasable_moves.mapped("picking_id.printed")):
+            return True
+        return unreleasable_moves.filtered(
+            lambda m: not m.rule_id.allow_unrelease_return_done_move and m.quantity_done
+        )
 
     def _is_unrelease_allowed_on_origin_moves(self, origin_moves):
         """We check that the origin moves are in a state that allows the unrelease
         of the current move. At this stage, a move can't be unreleased if
-          * a picking is already printed. (The work on the picking is planned and
-            we don't want to change it)
-          * a quantity done is recorded
-          * the processed origin moves is not consumed by the dest moves.
+        the processed origin moves is not consumed by the dest moves.
         """
         self.ensure_one()
+        if self.rule_id.allow_unrelease_return_done_move:
+            return True
         origin_done_moves = origin_moves.filtered(lambda m: m.state == "done")
```

Comment:
Because there is also the return_picking_type_id condition that must be valid
```suggestion
        origin_done_moves = origin_moves.filtered(lambda m: m.state == "done")
        if self.rule_id.allow_unrelease_return_done_move:
            origin_done_moves = origin_done_moves.filtered(lambda m: not m.picking_type_id.return_picking_type_id)
```